### PR TITLE
Add differentiability tests to conv layers

### PR DESCRIPTION
**Pull request by ilancoulon**
Those tests can show if the conv layers are differentiable or not. For now, it seems they're not -> the tests fail.

The main error of the tests is in the last message of #54 , rewritten here:
```
layer without graph: Error During Test at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\test\layers\conv.jl:249
  Got exception outside of a @test
  Mutating arrays is not supported
  Stacktrace:
   [1] error(::String) at .\error.jl:33
   [2] (::Zygote.var"#463#464")(::Nothing) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\lib\array.jl:71
   [3] (::Zygote.var"#1019#back#465"{Zygote.var"#463#464"})(::Nothing) at C:\Users\ilanc\.julia\packages\ZygoteRules\6nssF\src\adjoint.jl:49
   [4] update_batch_vertex at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\src\layers\msgpass.jl:47 [inlined]
   [5] (::typeof(∂(update_batch_vertex)))(::FillArrays.Fill{Float32,2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}}}) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface2.jl:0
   [6] propagate at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\src\layers\msgpass.jl:70 [inlined]
   [7] (::typeof(∂(propagate)))(::Nothing) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface2.jl:0
   [8] EdgeConv at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\src\layers\conv.jl:438 [inlined]
   [9] (::typeof(∂(λ)))(::Nothing) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface2.jl:0
   [10] #34 at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\test\layers\conv.jl:259 [inlined]
   [11] (::typeof(∂(λ)))(::Float32) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface2.jl:0
   [12] (::Zygote.var"#50#51"{Params,Zygote.Context,typeof(∂(λ))})(::Float32) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface.jl:177
   [13] gradient(::Function, ::Params) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface.jl:54
   [14] top-level scope at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\test\layers\conv.jl:259
   [15] top-level scope at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\Test\src\Test.jl:1113
   [16] top-level scope at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\test\layers\conv.jl:250
   [17] top-level scope at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\Test\src\Test.jl:1113
   [18] top-level scope at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\test\layers\conv.jl:236
   [19] top-level scope at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\Test\src\Test.jl:1113
   [20] top-level scope at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\test\layers\conv.jl:13
   [21] include(::String) at .\client.jl:439
   [22] macro expansion at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\test\runtests.jl:57 [inlined]
   [23] macro expansion at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\Test\src\Test.jl:1113 [inlined]
   [24] top-level scope at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\test\runtests.jl:56
   [25] include(::String) at .\client.jl:439
   [26] top-level scope at REPL[2]:1
   [27] eval(::Module, ::Any) at .\boot.jl:331
   [28] eval_user_input(::Any, ::REPL.REPLBackend) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:86
   [29] macro expansion at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\REPL\src\REPL.jl:118 [inlined]
```

### CODE REVIEW

**Comment by ilancoulon** [comment]
@yuehhua I merged your `gat` branch. I changed some functions and now the only problem we have left would be the ChebConv layer (all the conv differentiability tests pass but the ChebConv ones).

I'm not sure how the ChebConv function should be written to be automatically differentiable, I'll try to work on that.

**Comment by ilancoulon** [comment]
All the tests pass now. Nevertheless, there is still an issue. To write an adjoint for eigen, I used https://github.com/GiggleLiu/BackwardsLinalg.jl/blob/master/src/symeigen.jl
The issue is that it only works with symetric matrices, and therefore the `scaled_laplacian` function could only be used with undirect graphs. I added a symmetry check. 
I don't think having a eigenmax automatic differentiation for any matrix would be that simple.

**Comment by yuehhua** [comment]
Adjoint ofr eigen is provided by Zygote.jl: https://github.com/FluxML/Zygote.jl/blob/a6e8c4e9a904ddfbb8e2df78e459add74ae335d9/src/lib/array.jl#L572

**Comment by ilancoulon** [comment]
Indeed, trying that right away.

**Comment by ilancoulon** [comment]
It is weird, when I try to directly use eigen I get the following error lauching the tests:
```
layer without graph: Error During Test at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\test\layers\conv.jl:67
  Got exception outside of a @test
  Can't differentiate foreigncall expression
  Stacktrace:
   [1] error(::String) at .\error.jl:33
   [2] syevr! at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\LinearAlgebra\src\lapack.jl:5072 [inlined]
   [3] (::typeof(∂(syevr!)))(::Tuple{Array{Float64,1},Nothing}) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface2.jl:0
   [4] #eigen!#103 at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\LinearAlgebra\src\symmetric.jl:658 [inlined]
   [5] (::typeof(∂(#eigen!#103)))(::NamedTuple{(:values, :vectors),Tuple{Array{Float64,1},Nothing}}) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface2.jl:0 (repeats 2 times)
   [6] #eigen!#75 at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\LinearAlgebra\src\eigen.jl:151 [inlined]
   [7] (::typeof(∂(#eigen!#75)))(::NamedTuple{(:values, :vectors),Tuple{Array{Float64,1},Nothing}}) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface2.jl:0 (repeats 2 times)
   [8] #eigen#77 at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\LinearAlgebra\src\eigen.jl:237 [inlined]
   [9] (::typeof(∂(#eigen#77)))(::NamedTuple{(:values, :vectors),Tuple{Array{Float64,1},Nothing}}) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface2.jl:0
   [10] eigen at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.4\LinearAlgebra\src\eigen.jl:235 [inlined]
   [11] (::typeof(∂(eigen)))(::NamedTuple{(:values, :vectors),Tuple{Array{Float64,1},Nothing}}) at C:\Users\ilanc\.julia\packages\Zygote\1GXzF\src\compiler\interface2.jl:0
   [12] scaled_laplacian at C:\Users\ilanc\Dropbox\Stage-MTL\julia\GeometricFlux.jl\src\operations\linalg.jl:143 [inlined]
```

**Comment by yuehhua** [comment]
It needs to be symmetric.

```
julia> using LinearAlgebra

julia> A = rand(5,5)
5×5 Array{Float64,2}:
 0.340832  0.442944  0.00733532  0.180204   0.490136
 0.859511  0.490116  0.792274    0.0858942  0.165364
 0.38924   0.407874  0.0461607   0.878398   0.165644
 0.448136  0.505685  0.651879    0.659611   0.779182
 0.804118  0.670956  0.256478    0.415166   0.836159

julia> A = A + A'
5×5 Array{Float64,2}:
 0.681665  1.30246   0.396575   0.62834   1.29425
 1.30246   0.980233  1.20015    0.591579  0.83632
 0.396575  1.20015   0.0923213  1.53028   0.422122
 0.62834   0.591579  1.53028    1.31922   1.19435
 1.29425   0.83632   0.422122   1.19435   1.67232
julia> A = LinearAlgebra.Symmetric(A)
5×5 Symmetric{Float64,Array{Float64,2}}:
 0.681665  1.30246   0.396575   0.62834   1.29425
 1.30246   0.980233  1.20015    0.591579  0.83632
 0.396575  1.20015   0.0923213  1.53028   0.422122
 0.62834   0.591579  1.53028    1.31922   1.19435
 1.29425   0.83632   0.422122   1.19435   1.67232

julia> gradient(x -> sum(eigen(x).values), A)
([0.9999999999999997 -6.142422466815469e-16 … 3.624532685652299e-16 3.3553967186913047e-16; -6.142422466815469e-16 0.9999999999999999 … 7.141510434765108e-16 -4.1778963736823646e-16; … ; 3.624532685652299e-16 7.141510434765108e-16 … 0.999999999999999 -2.6844946814188936e-16; 3.3553967186913047e-16 -4.1778963736823646e-16 … -2.6844946814188936e-16 1.0000000000000007],)
```

**Comment by yuehhua** [comment]
```
julia> gradient(x -> maximum(eigen(x).values), A)
([0.17060176125642307 0.18265799908623612 … 0.20046697393223664 0.21614155886028366; 0.18265799908623612 0.19556623791263056 … 0.21463375331921664 0.23141604383239003; … ; 0.20046697393223664 0.21463375331921664 … 0.2355603326811201 0.253978879975286; 0.21614155886028366 0.23141604383239003 … 0.253978879975286 0.2738375801193236],)
```

**Comment by ilancoulon** [comment]
This passed all the tests!

**Comment by yuehhua** [comment]
I have two commits on my gat branch. I need the softmax one.
I have no idea why I just cannot push to this branch.......

**Comment by yuehhua** [comment]
Could you help me with that?

**Comment by ilancoulon** [comment]
You should be able to add commits (I had already set the option). You need to push on my own repository though, so you would need to add it as a remote. Some commands like those should work:
```shell
git remote add icfork https://github.com/ilancoulon/GeometricFlux.jl
git fetch icfork
git checkout icfork/test_conv_diff
git merge gat
git push icfork/test_conv_diff
```

**Comment by ilancoulon** [comment]
Or I can do it for you!

**Comment by yuehhua** [comment]
OK, now it works.

File: test/layers/conv.jl [hunk 1]
```diff
@@ -20,6 +20,9 @@ adj = [0. 1. 0. 1.;
 
             Y = gc(X)
             @test size(Y) == (out_channel, N)
+            # Test that the gradient can be computed
+            Zygote.gradient(() -> sum(gc(X)), params(gc))
```
Comment (yuehhua) [inline test/layers/conv.jl:24]: To check the gradient, you could reference [gradcheck](https://github.com/FluxML/Zygote.jl/blob/master/test/gradcheck.jl) here.
I think the gradient test can be isolated as an another file `test/layer/grad.jl`.

Comment (ilancoulon) [reply test/layers/conv.jl:24]: Oh I didn't know the gradtest function, very useful thank you!

Comment (ilancoulon) [reply test/layers/conv.jl:24]: It seems to be in their `test` directory, therefore not available in the module, do you think we should copy the functions?

Comment (ilancoulon) [reply test/layers/conv.jl:24]: And unfortunately the gradcheck function only works with arrays as parameters. However, our params are of Flux' type `Params` and don't seem to work with that function. Furthermore, I think using ngradient on (kinda) complicated layers like the messagepassing ones would take too much time.

Comment (yuehhua) [reply test/layers/conv.jl:24]: I think gradcheck can be modified slightly to handle `Params`. It's not that hard though. Could you do this?

Comment (ilancoulon) [reply test/layers/conv.jl:24]: I'd love to but I'd say I'm not sure how I would do it. Could you point me in the right direction please?

**Comment by ilancoulon** [comment]
We can close it in favor of #102

**Pull request closed without merging**